### PR TITLE
Storybook: Add StoryShots integration to generate unit tests

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Part of #17973.
Depends on #18030.

> StoryShots adds automatic Jest Snapshot Testing for Storybook.

I followed the docs at this page: 
https://github.com/storybookjs/storybook/tree/master/addons/storyshots/storyshots-core

## Testing

Run `npm run test-unit packages/components/storybook/`

## Known issue

~There is a known issue with using React hooks with components in stories:
https://github.com/storybookjs/storybook/issues/8177~

~I'm seeing the following:~

> ```bash
>   ● @wordpress/components › Checkbox Control › Default
> 
>     Invariant Violation: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
>     1. You might have mismatching versions of React and the renderer (such as React DOM)
>     2. You might be breaking the Rules of Hooks
>     3. You might have more than one copy of React in the same app
>     See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
> 
>       17 | 
>       18 | export const _default = () => {
>     > 19 |      const [ isChecked, setChecked ] = useState( true );
>          |                                        ^
>       20 |      return (
>       21 |              <CheckboxControl
>       22 |                      label="Is author"
> ```

This issue was resolved by refactoring https://github.com/WordPress/gutenberg/pull/18030. This PR was rebased with `update/storybook-enhancements` to solve it here as well.

### CODE REVIEW

File: packages/components/storybook/config.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ addDecorator( withA11y );
 addDecorator( withKnobs );
 configure(
 	[
-		require.context( '../docs', true, /\/.+\.mdx$/ ),
+		// require.context( '../docs', true, /\/.+\.mdx$/ ),
```
Comment (gziolo) [inline packages/components/storybook/config.js]: I didn't investigate what to do about `mdx`. The simplest approach would  load only files with the `.js` extension using:
https://github.com/storybookjs/storybook/tree/master/addons/storyshots/storyshots-core#storykindrege

Comment (epiqueras) [reply packages/components/storybook/config.js]: Yeah, let's do that. Most of these addons don't have support for looking for code snippets inside `.mdx` files unfortunately.

Comment (gziolo) [reply packages/components/storybook/config.js]: I resolved it a bit differently, by loading `MDX` files only in non-test env.

**Comment by gziolo** [comment]
I figured out that some tests can't work in this setup. `ClipboardButton` is one of them:
Screen Shot 2019-10-21 at 14 38 54

We can skip such stories using the convention. In this case, it would be `DontTest` keyword in the title.  It shows up in the UI at the moment, but it should be easy to introduce a helper function which appends this pattern only in the `test` env.

**Comment by gziolo** [comment]
cc @diegohaz and @ItsJonQ who recently spent some time on writing unit tests for UI components. I'd love to hear your feedback on the presented approach.

File: packages/components/storybook/config.js (hunk 1 above)
Comment (epiqueras) [reply packages/components/storybook/config.js]: Nice!

**Comment by epiqueras** [comment]
>I figured out that some tests can't work in this setup. ClipboardButton is one of them:

What makes it break? Is there anyway we can mock it?

>In this case, it would be DontTest keyword in the title.

We can also just add an explicit list of `(?!)` to the `storyKindRegex` since the list will probably be somehow limited.

**Comment by gziolo** [comment]
> > I figured out that some tests can't work in this setup. ClipboardButton is one of them:
> 
> What makes it break? Is there anyway we can mock it?
> 
> > In this case, it would be DontTest keyword in the title.
> 
> We can also just add an explicit list of `(?!)` to the `storyKindRegex` since the list will probably be somehow limited.
> > I figured out that some tests can't work in this setup. ClipboardButton is one of them:

This is the full error:

```bash
  ● @wordpress/components › ClipboardButton › Default

    TypeError: Cannot read property 'firstChild' of null

      27 |      componentDidMount() {
      28 |              const { container, getText, onCopy } = this;
    > 29 |              const button = container.firstChild;
         |                                       ^
      30 | 
      31 |              this.clipboard = new Clipboard( button, {
      32 |                      text: getText,

      at ClipboardButton.firstChild [as componentDidMount] (packages/components/src/clipboard-button/index.js:29:28)
      at commitLifeCycles (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10930:22)
      at commitLayoutEffects (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13583:7)
      at HTMLUnknownElement.callCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10487:14)
      at invokeEventListeners (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:193:27)
      at HTMLUnknownElementImpl._dispatch (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
      at HTMLUnknownElement.dispatchEvent (node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
      at Object.invokeGuardedCallbackDev (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10537:16)
      at invokeGuardedCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10590:31)
      at commitRootImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13355:9)
      at unstable_runWithPriority (node_modules/scheduler/cjs/scheduler.development.js:643:12)
      at runWithPriority (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1887:10)
      at commitRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13184:3)
      at runRootCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12420:20)
      at node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1935:24
      at unstable_runWithPriority (node_modules/scheduler/cjs/scheduler.development.js:643:12)
      at runWithPriority (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1887:10)
      at flushSyncCallbackQueueImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1931:7)
      at flushSyncCallbackQueue (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1920:3)
      at scheduleUpdateOnFiber (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12297:9)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15087:3)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15115:10)
      at updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15134:10)
      at create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15824:5)
      at getRenderedTree (node_modules/@storybook/addon-storyshots/dist/frameworks/react/renderTree.js:24:16)
      at node_modules/@storybook/addon-storyshots/dist/test-bodies.js:21:18
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/api/snapshotsTestsTemplate.js:44:33)
```


> What makes it break? Is there anyway we can mock it?

I guess there is something going on with how it gets mounted. I don't think it's so much important and we can skip it but feel free to investigate further.
 
> > In this case, it would be DontTest keyword in the title.
> 
> We can also just add an explicit list of `(?!)` to the `storyKindRegex` since the list will probably be somehow limited.

We can start with the blacklist at first. Good idea 👍

**Comment by epiqueras** [comment]
Sounds good, let's do it.

**Comment by gziolo** [comment]
With 7c6e7b3 we have a list of skipped components in the Storyshots configuration. It should be ready to go. This branch is based on the #18030 so we can merge it there once approved.

**Comment by gziolo** [comment]
> With [7c6e7b3](https://github.com/WordPress/gutenberg/commit/7c6e7b351273c24a68f5ffe7b97298f038e7b9f4) we have a list of skipped components in the Storyshots configuration. It should be ready to go. This branch is based on the #18030 so we can merge it there once approved.

Now that #18030 landed, this branch was rebased against `master` and updated with the latest changes.

**Comment by gziolo** [comment]
@getdave @nerrad @diegohaz @jameslnewell - I saw your comment about testing utils in #17249. What's your take on the integration of StoryShots?

**Comment by nerrad** [comment]
Since our use of storybook is to create stories that provide examples of what to expect with component usage (and thus the established contracts on the provided prop interface), then I think anything that automates testing those expectations as well is a ➕from me (especially for components living in `@wordpress/components`).  It seems that we get some built-in assertions for free by virtue of simply writing stories, and then adding the relevant boilerplate to add the stories to the storyshots tests?

The _only_ concern I have would be if we rely on this as the sole mechanism for test coverage (which I don't expect we are).

**Comment by jameslnewell** [comment]
While I think snapshots have a place, I’m always wary of them because they’re so easily skipped over and accepted by developers without digging into what’s changed. My feelings are similar to @nerrad, that any *additional* type of testing is a plus but I wonder if visual regression testing might provide more value. e.g. https://github.com/percy/percy-storybook

**Comment by nerrad** [comment]
> but I wonder if visual regression testing might provide more value

I like that idea, but by virtue of the snapshot tests being connected with actual _stories_, do we already get that for free? When a snapshot fails, I would expect whoever is working on things would jump into the storybook to examine that related story and figure if what changed was expected or not?

**Comment by jameslnewell** [comment]
With visual regression testing you get side-by-side images depicting before and after states with the relevant change highlighted, whereas if you open the storybook you only see the after state and the change might not be immediately obvious (you can see the before state by using git but it’s still not a side by side comparison). 

Visual regression testing is similar to snapshotting in that it is still a manual process that a dev has to review and accept but looking at images makes it easier to interpret visual changes than a large JSON diff!

**Comment by nerrad** [comment]
What's the overhead on visual regression testing in terms of time added to tests?

**Comment by nerrad** [comment]
I'm aware of what visual regression testing is, and the downside to comparing snapshots is noted. 
 I was mostly remarking that snapshots fails would provide a signal that something is wrong and then the story could be checked to see if the snapshot fail is expected considering what is being worked on (granted it would require awareness of what should be happening in the story 😄 )

To the point you made, I think visual regression testing is really good for general app views consisting of a lot of composed components where ya a snapshot diff would be much more intimidating to review, but I'm not sure they add that much more value for more atomic components?

**Comment by jameslnewell** [comment]
> I'm aware of what visual regression testing is, and the downside to comparing snapshots is noted.

Sorry if I came across that way!!! I assumed you knew, I was only attempting to highlight that particular difference. 

> I'm not sure they add that much more value for more atomic components?

What are we testing or expecting to pick up from snapshotting?

> What's the overhead on visual regression testing in terms of time added to tests?

A totally uneducated guess would be 5 minutes. Probably only running on “pull_request”.

**Comment by gziolo** [comment]
> It seems that we get some built-in assertions for free by virtue of simply writing stories, and then adding the relevant boilerplate to add the stories to the storyshots tests?

There is no boilerplate at all. It works out of the box and gets (re)generated, the next time you run `npm run test-unit`. That's why I think it is so nice.

> The _only_ concern I have would be if we rely on this as the sole mechanism for test coverage (which I don't expect we are).

> While I think snapshots have a place, I’m always wary of them because they’re so easily skipped over and accepted by developers without digging into what’s changed. 

@nerrad and @jameslnewell No, we don't care about the test coverage at all. We removed all integrations which report the coverage in the past because we believe that numbers don't align with the quality of the test written. You can have 100% coverage which provides a very low-quality testing approach. Snapshots testing, in my opinion, falls under this group as well. However, they are easy to write and update. At the same time, all changes are stored in the revision control so you can use git history to figure out where the regression was introduced when the code review process failed. From my experience, contributors usually don't run unit tests locally, so this gets caught by Travis and this brings some focus on snapshots produced in the review process. We also have some examples where this type of testing is extremely helpful. See our own implementation of snapshot testing for block parsing and serialization.

> My feelings are similar to @nerrad, that any additional type of testing is a plus but I wonder if visual regression testing might provide more value. e.g. https://github.com/percy/percy-storybook

We would love to have visual testing integrated as well. Are there any alternative solutions that don't require external services? I see that Percy requires an API key and it has a quota for the basic usage. I added an item to the todo task in #17973.

> A totally uneducated guess would be 5 minutes. Probably only running on “pull_request”.

I'm sure it's way much slower than unit tests. It's similar to unit vs e2e tests. When we combine both together, then you can maximize the outcome having lower test coverage if you would analyze them independently. Just my way of thinking about all types of tests.

To conclude, I don't feel strongly about enforcing StoryShots as the have some downsides as discussed above. I'm fully aware of their limitations and I anticipate some struggles with their integration as already prooved with the `ClipboardButton` component. However, I can see the cases where they can help to catch regressions caused by changes in class names or by introducing some unexpected wrapper HTML tags. We observed issues like that in the past and this is where snapshot testing is helpful in the project with requirements like WordPress.

**Comment by gziolo** [comment]
> * that `a11y` hasn't been impacted? would other tooling like [`eslint-plugin-jsx-a11y`](https://www.npmjs.com/package/eslint-plugin-jsx-a11y) be more effective than a diff of markup?

We have it integrated for as long as I remember. I don't think on its own it was very helpful to catch more nuanced a11y issues.

**Review by epiqueras — CHANGES_REQUESTED** [review]
I left 2 suggestions I think we should implement and then I think we should merge this.

## On Snapshots vs. Screenshots

Screenshot testing is a nice-to-have when using paid services with smart diff heuristics, but most of the time it adds unnecessary noise and fails with the most minor of changes. The benefits are marginal if any. If a component has been touched in a PR, developers are expected to see what has changed visually before approving it and they will do a much better job at judging whether the changes are acceptable or not than a computer.

Snapshot testing does add a lot of value to projects like this though. WP cares a lot about markup and class names, because other code relies on it. Sometimes PRs might slip through where visually there weren't any significant changes. And, even when inspecting the code, things looked fine. But, simple logic changes can change markup and class names in specific cases and break things that relied on them. A Storyshot would have caught those changes and surfaced them to reviewers for them to judge whether they were acceptable or not.

In short, visual regression testing is something we already do manually in every PR and that we will still have to do even if a computer does it too, but DOM snapshot testing is something we don't do (We look at the declarative React code that declares all possible state outputs, but that makes it hard to catch breaking changes.) and that we would benefit from having a computer do.

## On a11y

We already use the Storybook [a11y plugin](https://github.com/storybookjs/storybook/tree/master/addons/a11y) which uses `axe` under the hood, which we can easily [integrate into the Storyshots](https://github.com/storybookjs/storybook/issues/3526).

File: babel.config.js [hunk 2]
```diff
@@ -4,5 +4,10 @@ module.exports = function( api ) {
 	return {
 		presets: [ '@wordpress/babel-preset-default' ],
 		plugins: [ 'babel-plugin-inline-json-import' ],
+		env: {
+			test: {
```
[the remark this answers is not in the document]
Comment (gziolo) [reply babel.config.js]: It's already scoped to stories with the current setup:
- it is scoped only for Storybook because the Babel config is local for the tool
- it gets activated only when running tests

I don't think there is any difference between those two options and they also don't share any preference in their docs. In addition to that, macros struggle with the cache invalidation for the dependent files. See the comment in `create-react-app` https://github.com/facebook/create-react-app/commit/11737bc78676868e76ce5d0c04ddbb5c758d3908#diff-b1524f1e1ccabdb327a12f23b432b20bR13: 

> 
>     // Babel Macros are notoriously hard to cache, so they shouldn't be
>     // https://github.com/babel/babel/issues/8497
>     // We naively detect macros using their package suffix and insert a random
>     // caller name, a valid option accepted by Babel, to compose a one-time
>     // cacheIdentifier for the file. We cannot tune the loader options on a per
>     // file basis.

File: packages/components/storybook/test/index.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import path from 'path';
+
+/**
+ * The list of components that should be skipped because they
+ * don't work with the default Storyshots setup.
+ *
+ * @type {string[]}
+ */
+const skippedComponents = [
+	'ClipboardButton',
```
[the remark this answers is not in the document]
Comment (gziolo) [reply packages/components/storybook/test/index.js]: This raises a few questions.
- do we want to mock anything on the case by case basis if something doesn't work out of the box?
- what's the goal of snapshot testing in this context? 
- are we sure that there isn't an issue with the current implementation of the `ClipboardButton` components?

I don't feel strongly about having 100% coverage through stories. I find StoryShots a nice tool that helps us to prevent regressions but not necessarily something which is mandatory for each component.

**Comment by gziolo** [comment]
> ## On a11y
> We already use the Storybook [a11y plugin](https://github.com/storybookjs/storybook/tree/master/addons/a11y) which uses `axe` under the hood, which we can easily [integrate into the Storyshots](https://github.com/storybookjs/storybook/issues/3526).

That's cool, we should definitely explore that as a follow-up task. I'd love to see integrated with the default configuration. Thanks for sharing 💯

File: babel.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (epiqueras) [reply babel.config.js]: That makes sense, thanks for clarifying.

File: packages/components/storybook/test/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (epiqueras) [reply packages/components/storybook/test/index.js]: >do we want to mock anything on the case by case basis if something doesn't work out of the box?

It's not on a case by case basis, refs can be mocked generically, but yeah, why not?

>what's the goal of snapshot testing in this context?

To make sure the classes and markup of the button don't change in a breaking way.

>are we sure that there isn't an issue with the current implementation of the ClipboardButton components?

Refs are just not supported and we need to mock them.

[the remark this answers is not in the document]
Comment (jameslnewell) [reply packages/components/storybook/test/index.js]: > To make sure the classes and markup of the button don't change in a breaking way.

This is where I'm a little suspect at the value of snapshot testing. I'm assuming that any logic in a component that creates a meaningful markup change/variation (e.g. classnames, a11y attributes etc) would be tested by regular unit tests (is that correct?) and if that assumption is true then it doesn't feel like snapshot tests would provide much value over a `git diff` of the component for a developer to be aware of whether hard-coded markup and attributes have changed?

Is there something else I'm missing? 🤔

[the remark this answers is not in the document]
Comment (epiqueras) [reply packages/components/storybook/test/index.js]: >I'm assuming that any logic in a component that creates a meaningful markup change/variation (e.g. classnames, a11y attributes etc) would be tested by regular unit tests (is that correct?)

That is rarely the case, and snapshot testing makes this easier anyways.

[the remark this answers is not in the document]
Comment (jameslnewell) [reply packages/components/storybook/test/index.js]: Thanks for clarifying @epiqueras!

**Comment by ItsJonQ** [comment]
> I'd love to hear your feedback on the presented approach.
@gziolo Hallo! Apologies for the delay!

From my experience, snapshot testing (in general) is definitely better than nothing, but I prefer having and writing unit/integration tests.

I also think writing unit tests should also be easier, which can be achieved by having testing utils (if we're staying with Enzyme) or leveraging another testing solution.

Personally, I haven't used this particular addon for Storybook.

I'm okay with keeping this in to see where it goes/how it feels :)

Maybe it can automate a portion of manually writing basic render tests, freeing up time/focus to write unit/integration tests for more complex scenarios.

File: storybook/config.js [hunk 4]
```diff
@@ -16,8 +16,10 @@ addDecorator( withA11y );
 addDecorator( withKnobs );
 configure(
 	[
-		require.context( './stories/', true, /\/.+\.(js|mdx)$/ ),
+		// StoryShots addon doesn't support MDX files at the moment.
+		// It should ignore the playground in the initial pass as well.
+		process.env.NODE_ENV !== 'test' && require.context( './stories/', true, /\/.+\.(js|mdx)$/ ),
```
Comment (jorgefilipecosta) [inline storybook/config.js]: Could not calling require.context for test NODE_ENV have another undesired effect? Would it make sense to in the test environment call require.context but not include mdx in the regex?

**Review by jorgefilipecosta — APPROVED** [review]
I was able to run the tests with success by using "npm run test-unit storybook/test/index.js".

I'm not sure if these tests will be useful in catching regressions or not but I think the best way is to merge them and check if they helped us or they added developer overhead to update the snapshots.
We also have some interesting potential follow-ups pending like the a11y tests, so I guess we should merge this PR and see where these type of tests leaves us.

File: storybook/test/index.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import path from 'path';
+
+/**
+ * The list of components that should be skipped because they
+ * don't work with the default Storyshots setup.
+ *
+ * @type {string[]}
+ */
+const skippedComponents = [
+	'ClipboardButton',
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply storybook/test/index.js]: https://github.com/WordPress/gutenberg/pull/18031#discussion_r339701756

File: storybook/config.js (hunk 4 above)
Comment (epiqueras) [reply storybook/config.js]: The issue is that we are calling it for JS files twice now. See line 20/22.

Comment (gziolo) [reply storybook/config.js]: There are two different entry paths defined in line 20:
```js
require.context( './stories/', true, /\/.+\.(js|mdx)$/ )
```
and line 22:
```js
require.context( '../packages/components/src/', true, /\/stories\/.+\.js$/ )
```

> Could remove the call require.context for test NODE_ENV have an undesired effect? Would it make sense to in the test environment call require.context but not include mdx in the regex?

It depends if we want to enable any stories in the `storybook/stories` folder. At the moment it contains MDX file and the playground. We can always split MDX to its own line and let Storyshot run the test again playground. It might work, I didn't check.

File: storybook/test/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gziolo) [reply storybook/test/index.js]: Yes, it explodes and this is the full report:
```
    TypeError: Cannot read property 'firstChild' of null

      27 |      componentDidMount() {
      28 |              const { container, getText, onCopy } = this;
    > 29 |              const button = container.firstChild;
         |                                       ^
      30 | 
      31 |              this.clipboard = new Clipboard( button, {
      32 |                      text: getText,

      at ClipboardButton.firstChild [as componentDidMount] (packages/components/src/clipboard-button/index.js:29:28)
      at commitLifeCycles (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10930:22)
      at commitLayoutEffects (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13583:7)
      at HTMLUnknownElement.callCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10487:14)
      at invokeEventListeners (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:193:27)
      at HTMLUnknownElementImpl._dispatch (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
      at HTMLUnknownElement.dispatchEvent (node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
      at Object.invokeGuardedCallbackDev (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10537:16)
      at invokeGuardedCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:10590:31)
      at commitRootImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13355:9)
      at unstable_runWithPriority (node_modules/scheduler/cjs/scheduler.development.js:643:12)
      at runWithPriority (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1887:10)
      at commitRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13184:3)
      at runRootCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12420:20)
      at node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1935:24
      at unstable_runWithPriority (node_modules/scheduler/cjs/scheduler.development.js:643:12)
      at runWithPriority (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1887:10)
      at flushSyncCallbackQueueImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1931:7)
      at flushSyncCallbackQueue (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1920:3)
      at scheduleUpdateOnFiber (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12297:9)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15087:3)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15115:10)
      at updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15134:10)
      at create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15824:5)
      at getRenderedTree (node_modules/@storybook/addon-storyshots/dist/frameworks/react/renderTree.js:24:16)
      at node_modules/@storybook/addon-storyshots/dist/test-bodies.js:21:18
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/api/snapshotsTestsTemplate.js:44:33)
```

I didn't want to spend too much time on this case as this is a simple button to display with advanced logic :)

**Comment by gziolo** [comment]
I invested more time into this issue and I managed to mock those nodes which use refs with the following commits:
1a59025
ccc705b

This means, we no longer skip any stories from snapshot testing. However, the side effect is that if there is some special use case for refs, it might require custom logic in the mocking helper. The following seem to work for now:

```js
		createNodeMock: ( element ) => {
			if ( story.kind === 'Components|ClipboardButton' ) {
				return {
					firstChild: document.createElement( 'button' ),
				};
			}
			return element.type && document.createElement( element.type );
		},
```

I followed links shared by @epiqueras:
- https://reactjs.org/blog/2016/11/16/react-v15.4.0.html#mocking-refs-for-snapshot-testing
- https://github.com/storybookjs/storybook/tree/master/addons/storyshots/storyshots-core#using-createnodemock-to-mock-refs

It looks like it's a general issue with snapshot testing in React apps.

File: storybook/config.js (hunk 4 above)
Comment (epiqueras) [reply storybook/config.js]: Oh right, I missed that :sweat_smile:

Comment (epiqueras) [reply storybook/config.js]: I think it's fine as is for now then.

**Comment by diegohaz** [comment]
One thing I noticed is that storyshots tests will still fail if a console warning is emitted. On #17875, I created a story for a deprecated thing so I could see the UI worked and the warning was shown.

On normal tests I can check if the warning was shown. Can I do that with Storyshots?

**Review by epiqueras — APPROVED** [review]

**Comment by gziolo** [comment]
> One thing I noticed is that storyshot tests will still fail if a console warning is emitted. On #17875, I created a story for a deprecated thing so I could see the UI worked and the warning was shown.

This raises the question of whether we should keep the legacy version of Toolbar in Storybook 🤔 

I could make the point that it should be removed from there as it should be discouraged from using in new code. Having it included in the docs might create the temptation for folks to copy the example and use it in their projects when they find that it fits their needs.

I'm aware that it isn't probably the type of answers you expected but this is something we should discuss more broadly before we jump into finding a technical solution on how to mitigate it. Maybe we don't have to do it. The only downside is that this story is very useful in the process of developing the migration path for the `Toolbar` component.

**Comment by gziolo** [comment]
I included documentation in 92f46dd.
